File: src/Feature Extraction/pc_ext.cpp
#include <pcl/io/pcd_io.h>
#include <pcl/point_types.h>
#include <pcl/features/normal_3d.h>
#include <pcl/features/moment_of_inertia_estimation.h>
#include <iostream>

int main(int argc, char** argv)
{
  // .pcd 파일 경로 설정
  std::string file_path = "/media/rtlink/JetsonSSD-256/socap_dataset/pc/1026_1581624733.968697.pcd";

  // .pcd 파일에서 포인트 클라우드 읽기
  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
  if (pcl::io::loadPCDFile<pcl::PointXYZ>(file_path, *cloud) == -1) // 파일 로딩 실패
  {
    PCL_ERROR("Failed to read .pcd file\n");
    return (-1);
  }

  // Normal 추정
  pcl::NormalEstimation<pcl::PointXYZ, pcl::Normal> ne;
  ne.setInputCloud(cloud);

  pcl::search::KdTree<pcl::PointXYZ>::Ptr tree(new pcl::search::KdTree<pcl::PointXYZ>());
  ne.setSearchMethod(tree);

  pcl::PointCloud<pcl::Normal>::Ptr normals(new pcl::PointCloud<pcl::Normal>);
  ne.setRadiusSearch(0.03); // Normal 추정 반경 설정
  ne.compute(*normals);

  // Moment of Inertia 추정
  pcl::MomentOfInertiaEstimation<pcl::PointXYZ> feature_extractor;
  feature_extractor.setInputCloud(cloud);
  feature_extractor.setInputNormals(normals);

  pcl::PointXYZ min_point, max_point;
  pcl::PointXYZ center_point;
  Eigen::Matrix3f moment_of_inertia;
  std::vector<float> radii;

  feature_extractor.compute();
  feature_extractor.getAABB(min_point, max_point);
  feature_extractor.getEigenVectors(moment_of_inertia);
  feature_extractor.getEigenValues(radii);
  feature_extractor.getMassCenter(center_point);

  // Surface Roughness 계산
  float roughness = 0;
  for (size_t i = 0; i < cloud->points.size(); ++i)
  {
    Eigen::Vector3f vec(cloud->points[i].x - center_point.x, cloud->points[i].y - center_point.y, cloud->points[i].z - center_point.z);
    vec = moment_of_inertia * vec;
    float dist = vec.norm();
    roughness += dist;
  }
  roughness /= cloud->points.size();

  // Surface Roughness 출력
  std::cout << "Surface Roughness: " << roughness << std::endl;

  return 0;
}

File: src/Feature Extraction/pc_ext.py
import open3d as o3d
import numpy as np

# .pcd 파일 경로 설정
file_path = "/media/rtlink/JetsonSSD-256/socap_dataset/pc/1026_1581624733.968697.pcd"

# .pcd 파일에서 포인트 클라우드 읽기
pcd = o3d.io.read_point_cloud(file_path)

# Normal 벡터 추정
pcd.estimate_normals(search_param=o3d.geometry.KDTreeSearchParamHybrid(radius=0.03, max_nn=30))

# Moment of Inertia 계산
min_bound, max_bound = pcd.get_axis_aligned_bounding_box()
center = (min_bound + max_bound) / 2.0
R = pcd.get_rotation_matrix_from_xyz((np.pi / 2, 0, np.pi / 2))
pcd.rotate(R, center=center)
moi = o3d.geometry.MomentOfInertiaEstimation()
moi.estimate(pcd)
moment_of_inertia = moi.get_inertia_tensor()
evals, evecs = np.linalg.eig(moment_of_inertia)
radii = np.sqrt(evals)

# Surface Roughness 계산
roughness = 0
for i in range(len(pcd.points)):
    point = np.array(pcd.points[i])
    vec = point - center
    dist = np.linalg.norm(np.dot(vec, evecs) / radii)
    roughness += dist
roughness /= len(pcd.points)

# Surface Roughness 출력
print("Surface Roughness:", roughness)
